ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/upnp/UpnpNatManager.java

```diff
@@ -0,0 +1,487 @@
+              PortMappingAdd callback =
+                  new PortMappingAdd(service, portMapping) {
+                    /**
+                     * Because the underlying jupnp library omits generics info in this method
+                     * signature, we must too when we override it.
+                     */
+                    @Override
+                    @SuppressWarnings("rawtypes")
+                    public void success(final ActionInvocation invocation) {
+                      LOG.info(
+                          "Port forward request for "
+                              + portMapping.getProtocol()
+                              + " "
+                              + portMapping.getInternalPort()
+                              + " -> "
+                              + portMapping.getExternalPort()
+                              + " succeeded");
+                      upnpQueryFuture.complete(null);
+                    }
+
+                    /**
+                     * Because the underlying jupnp library omits generics info in this method
+                     * signature, we must too when we override it.
+                     */
+                    @Override
+                    @SuppressWarnings("rawtypes")
+                    public void failure(
+                        final ActionInvocation invocation,
+                        final UpnpResponse operation,
+                        final String msg) {
+                      LOG.warn(
+                          "Port forward request for "
+                              + portMapping.getProtocol()
+                              + " "
+                              + portMapping.getInternalPort()
+                              + " -> "
+                              + portMapping.getExternalPort()
+                              + " failed");
+                      upnpQueryFuture.completeExceptionally(new Exception(msg));
+                    }
+                  };
+
+              LOG.info(
+                  "Requesting port forward "
+                      + portMapping.getProtocol()
+                      + " "
+                      + portMapping.getInternalPort()
+                      + " -> "
+                      + portMapping.getExternalPort());
+
+              upnpService.getControlPoint().execute(callback);
+
+              return upnpQueryFuture;
+            });
+  }
+
+  /**
+   * Recursively crawls the given device to look for specific services.
+   *
+   * @param device is the device to inspect for desired services.
+   * @param serviceTypes is a set of service types to look for.
+   */
+  private void inspectDeviceRecursive(final RemoteDevice device, final Set<String> serviceTypes) {
+    for (RemoteService service : device.getServices()) {
+      String serviceType = service.getServiceType().getType();
+      if (serviceTypes.contains(serviceType)) {
+        // TODO: handle case where service is already "recognized" as this could lead to
```

Comment:
I'll give this some thought. In the real world, I think this would happen for a few reasons:

1. There really are > 1 IGDs on the network, perhaps routing to very different networks.
2. Some equipment (etc.) is malfunctioning
3. Some equipment (etc.) is acting maliciously

In all these cases, I think throwing an error is probably better than trying to guess which IGD is preferred. However, by the time we discover IGD#2, we've already tried to proceed with IGD#1, including successfully completing futures such that we've lost our chance to handle this gracefully.

On the other hand, the easy thing to do is ignore IGD#2.

Thoughts?

Reply:
Is there a way to wait for a round of search to finish before inspecting the discovered devices?  I'm leaning towards something like # 2: we could wait some period of time after we `start` to collect the list of discovered devices, then inspect them all recursively to find the services we want or error out.

Reply:
> Is there a way to wait for a round of search to finish before inspecting the discovered devices? I'm leaning towards something like # 2: we could wait some period of time after we `start` to collect the list of discovered devices, then inspect them all recursively to find the services we want or error out.

I'm OK with that. But to be clear, this comes at the expense of the normal/best case where an IGD responds virtually instantaneously. And this is also for a tiny corner case...

At present, there is a 60sec timeout for the overall process. Perhaps we add a smaller (1-5 sec or so) timeout that kicks in after the first IGD response is heard where we wait to hear from any other IGD.

Reply:
Fair point - maybe we should keep it simple for now and just log a warning if we encounter a duplicate service.